File: November/rotatearr.py
# LeetCode 189: Rotate Array
# Explanation:
# 1. To rotate array right by k steps, we can reverse parts of the array:
#    - Reverse the entire array
#    - Reverse the first k elements
#    - Reverse the remaining elements
# 2. This achieves the rotation in-place without extra space.
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def rotate(self, nums: list[int], k: int) -> None:
        n = len(nums)
        k %= n
        nums.reverse()
        nums[:k] = reversed(nums[:k])
        nums[k:] = reversed(nums[k:])

    def rotateSlicing(self, nums: list[int], k: int) -> None:
        n = len(nums)
        k %= n
        nums.reverse()
        nums[:k] = reversed(nums[:k])
        nums[k:] = reversed(nums[k:])

    def rotateOneLine(self, nums: list[int], k: int) -> None: nums[:] = nums[-(k%len(nums)):] + nums[:-(k%len(nums))]

File: November/rotatearr.js
// LeetCode 189: Rotate Array
// Explanation:
// 1. We can reverse the array in three steps (like Python version).
// 2. Helper reverse() swaps elements in-place.
// Time Complexity: O(n)
// Space Complexity: O(1)

var rotate = function(nums, k) {
    const n = nums.length;
    k = k % n;
    
    const reverse = (arr, start, end) => {
        while (start < end) {
            [arr[start], arr[end]] = [arr[end], arr[start]];
            start++; end--;
        }
    };
    
    reverse(nums, 0, n - 1);
    reverse(nums, 0, k - 1);
    reverse(nums, k, n - 1);
};

